### PR TITLE
Issue84 refactor

### PR DESCRIPTION
**Pull request by technetos**
This PR encompass all of what #87 covers making #87 somewhat optional to merge in.  

Refactored codebase structure to account for recent changes and maintain manageability, readability and sanity ;)

### CODE REVIEW

File: hatch/src/assets/tests/fixtures.rs [hunk 1]
```diff
@@ -1,8 +1,10 @@
-use project::{ Project, Arch, Target, BuildConfig, ProjectKind };
+use project::{ Project, ProjectKind };
+use project::build::{ Config, Target };
+use platform::arch::Arch;
 use std::path::PathBuf;
 
 pub fn project(kind: ProjectKind) -> Project {
-  let config = BuildConfig::new(kind,
+  let config = Config::new(kind,
                                 String::from("g++"),
```
Comment (DannyPeck) [inline hatch/src/assets/tests/fixtures.rs]: Can you fix the indention here?

File: hatch/src/assets/tests/tuprules.rs [hunk 2]
```diff
@@ -59,7 +61,7 @@ else
 endif
 PROJECT_LIB = $(PROJECT).$(EXTENSION)");
 
-  let config = BuildConfig::new(ProjectKind::Library(LibraryKind::Static),
+  let config = Config::new(ProjectKind::Library(LibraryKind::Static),
                                 String::from("g++"),
```
Comment (DannyPeck) [inline hatch/src/assets/tests/tuprules.rs]: Fix indention.

File: hatch/src/assets/tests/tuprules.rs [hunk 3]
```diff
@@ -110,7 +112,7 @@ else
...
```
Comment (DannyPeck) [inline hatch/src/assets/tests/tuprules.rs]: Fix indention.

File: hatch/src/assets/tests/tuprules.rs [hunk 4]
```diff
@@ -161,7 +163,7 @@ else
 endif
 PROJECT_LIB = $(PROJECT).$(EXTENSION)");
 
-  let config = BuildConfig::new(ProjectKind::Library(LibraryKind::Shared),
+  let config = Config::new(ProjectKind::Library(LibraryKind::Shared),
                                 String::from("g++"),
```
Comment (DannyPeck) [inline hatch/src/assets/tests/tuprules.rs]: Fix indention.

File: hatch/src/assets/tests/tuprules.rs [hunk 5]
```diff
@@ -213,7 +215,7 @@ else
...
```
Comment (DannyPeck) [inline hatch/src/assets/tests/tuprules.rs]: Fix indention.

File: hatch/src/assets/tests/tuprules.rs [hunk 6]
```diff
@@ -252,7 +254,7 @@ TEST_OBJ_FILES = $(TEST_TARGET)/*.o
 # includes the STATIC and SHARED variables for the target platform
 include @(TUP_PLATFORM).tup");
 
-  let config = BuildConfig::new(ProjectKind::Binary,
+  let config = Config::new(ProjectKind::Binary,
                                 String::from("g++"),
```
Comment (DannyPeck) [inline hatch/src/assets/tests/tuprules.rs]: Fix indention.

File: hatch/src/assets/tests/tuprules.rs [hunk 7]
```diff
@@ -294,7 +296,7 @@ TEST_OBJ_FILES = $(TEST_TARGET)/*.o
...
```
Comment (DannyPeck) [inline hatch/src/assets/tests/tuprules.rs]: Fix indention.

File: hatch/src/assets/tests/tuprules.rs [hunk 8]
```diff
@@ -334,7 +336,7 @@ TEST_OBJ_FILES = $(TEST_TARGET)/*.o
...
```
Comment (DannyPeck) [inline hatch/src/assets/tests/tuprules.rs]: Fix indention.

File: hatch/src/assets/tests/tuprules.rs [hunk 9]
```diff
@@ -375,7 +377,7 @@ TEST_OBJ_FILES = $(TEST_TARGET)/*.o
...
```
Comment (DannyPeck) [inline hatch/src/assets/tests/tuprules.rs]: Fix indention.

File: hatch/src/deps/mod.rs [hunk 10]
```diff
@@ -1,24 +1,24 @@
...
 use task;
-
-pub fn modules_path(base: &Path) -> PathBuf {
-  base.join("hatch_modules")
-}
-
-pub fn hatchfile_path(base: &Path) -> PathBuf {
-  base.join("Hatch.yml")
-}
+use self::dependency::Dependency;
```
Comment (DannyPeck) [inline hatch/src/deps/mod.rs:12]: the use of self is cool!

File: hatch/src/deps/mod.rs [hunk 11]
```diff
@@ -1,24 +1,24 @@
...
+use locations::hatchfile_path;
 
 pub fn clone_dep(url: &str, path: &Path) {
   Repository::clone(url, path);
 }
 
-fn walk(path: &Path, callback: &mut FnMut(&Path) -> HatchResult<bool>) -> HatchResult<()> {
+fn walk(path: &Path,
+        callback: &mut FnMut(&Path) -> HatchResult<bool>) -> HatchResult<()>
```
Comment (DannyPeck) [inline hatch/src/deps/mod.rs:20]: having this on two lines is weird to me.

File: hatch/src/deps/tests/clone_one_dep.rs [hunk 12]
```diff
@@ -0,0 +1,9 @@
+use deps;
+use project::Dependency;
+
+#[test]
+fn clone_one_dep() {
+  let registry = Path::new("TestProject/hatch_modules");
+
+
```
Comment (DannyPeck) [inline hatch/src/deps/tests/clone_one_dep.rs]: Remove extra new lines.

File: hatch/src/platform/arch.rs [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+impl fmt::Display for Arch {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    match *self {
+      Arch::X64 => write!(f, "x64"),
+      Arch::X86 => write!(f, "x86"),
+    }
+  }
+}
+
+impl Arch {
```
Comment (DannyPeck) [inline hatch/src/platform/arch.rs:22]: That is cool that you can add functions like this to an enum!

File: hatch/src/project/mod.rs [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    match *self {
+      ProjectKind::Library(LibraryKind::Shared) => write!(f, "shared-lib"),
+      ProjectKind::Library(LibraryKind::Static) => write!(f, "static-lib"),
+      ProjectKind::Binary => write!(f, "binary"),
+    }
+  }
+}
+
+
```
Comment (DannyPeck) [inline hatch/src/project/mod.rs]: Remove this extra new line :).

**Review by DannyPeck — CHANGES_REQUESTED** [review]
This is great!  Just a few things here and there but I am really liking this!  Also, if you just want to merge the other PR into this one and do just this one I am ok with that.

File: hatch/src/deps/mod.rs (hunk 11 above)
Comment (technetos) [reply hatch/src/deps/mod.rs:20]: Its really long otherwise

Comment (technetos) [reply hatch/src/deps/mod.rs:20]: This is gonna change a good amount when we incorporate more git2-rs features into the codebase.

**Review by DannyPeck — APPROVED** [review]

**Pull request merged**